### PR TITLE
Added a hook to eraDat to update the leap seconds.

### PR DESCRIPTION
**Pull request by jwoillez**
DO NOT MERGE - Pull Request to stimulate discussions with respect to #43.

I have tried to keep the changes to SOFA/ERFA to a bare minimum. The PR intercepts the `changes` and `NDAT` parameters in `eraDat`, and provides a way to override them: `eraUpdateLeapSeconds`. Memory management and production of the appropriate `eraLEAPSECOND` array is left to the caller.

Note: the SOFA/ERFA test suite passes with these changes.

### CODE REVIEW

File: src/dat.c [hunk 1]
```diff
@@ -1,5 +1,14 @@
 #include "erfa.h"
 
+static eraLEAPSECOND *changes = 0;
+static int NDAT = 0;
```
Comment (jwoillez) [inline src/dat.c]: `changes` and `NDAT` are moved outside the scope of the `eraDat()` function, in order to be accessible by `eraUpdateLeapSeconds`.

File: src/dat.c [hunk 2]
```diff
@@ -1,5 +1,14 @@
...
+
+void eraUpdateLeapSeconds(eraLEAPSECOND *leapseconds, int count)
+{
+   changes = leapseconds;
+   NDAT = count;
+}
+
```
Comment (jwoillez) [inline src/dat.c]: `eraUpdateLeapSeconds` can now be used to change the leap seconds array `changes` and size `NDAT`.

File: src/dat.c [hunk 3]
```diff
@@ -144,10 +153,7 @@ int eraDat(int iy, int im, int id, double fd, double *deltat )
    enum { NERA1 = (int) (sizeof drift / sizeof (double) / 2) };
 
 /* Dates and Delta(AT)s */
-   static const struct {
-      int iyear, month;
-      double delat;
-   } changes[] = {
+   static const eraLEAPSECOND _changes[] = {
```
Comment (jwoillez) [inline src/dat.c]: This only uses the globally defined `eraLEAPSECOND` structure, defined in `erfa.h`.

File: src/dat.c [hunk 4]
```diff
@@ -193,7 +199,13 @@ int eraDat(int iy, int im, int id, double fd, double *deltat )
    };
 
 /* Number of Delta(AT) changes */
-   enum { NDAT = (int) (sizeof changes / sizeof changes[0]) };
+   enum { _NDAT = (int) (sizeof _changes / sizeof _changes[0]) };
+
```
Comment (jwoillez) [inline src/dat.c]: Just like `changes`, `NDAT` is renamed to `_NDAT` to allow interception.

**Comment by olebole** [comment]
This PR makes it easier to change the leap seconds programmatically, but it needs the caller to be aware of that. This may be the case for astropy; however other programs (like DS9) may not do this. A decentralized approach moves the responsibility to the (many) clients, which is IMO not the right way.

File: src/dat.c [hunk 5]
```diff
@@ -193,7 +199,13 @@ int eraDat(int iy, int im, int id, double fd, double *deltat )
...
 
 /* Number of Delta(AT) changes */
-   enum { NDAT = (int) (sizeof changes / sizeof changes[0]) };
+   enum { _NDAT = (int) (sizeof _changes / sizeof _changes[0]) };
+   
+/* Initialise leap seconds if needed */
+   if ( NDAT == 0) {
+      NDAT = _NDAT;
+      changes = (eraLEAPSECOND *)_changes;
+   }
```
Comment (jwoillez) [inline src/dat.c]: The first time `eraDat` is called, and if eraUpdateLeapSeconds() has not been called before, `NDAT` and `changes` are made to point to the private and static `_NDAT` and `_changes`. This preserves the current behavior of SOFA/ERFA.

File: src/erfa.h [hunk 6]
```diff
@@ -337,6 +337,7 @@ int eraGd2gce(double a, double f,
 /* Astronomy/Timescales */
 int eraD2dtf(const char *scale, int ndp, double d1, double d2,
              int *iy, int *im, int *id, int ihmsf[4]);
+void eraUpdateLeapSeconds(eraLEAPSECOND *leapseconds, int count);
```
Comment (jwoillez) [inline src/erfa.h]: This is the new public interface to update the leap seconds.

File: src/erfam.h [hunk 7]
```diff
@@ -41,6 +41,12 @@ typedef struct {
    double pv[2][3];   /* barycentric PV of the body (au, au/day) */
 } eraLDBODY;
 
+/* Leap second definition */
+typedef struct {
+   int iyear, month;
+   double delat;
+} eraLEAPSECOND;
+
```
Comment (jwoillez) [inline src/erfam.h]: The new public leap second structure, that comes along with `eraUpdateLeapSeconds()`.

**Comment by jwoillez** [comment]
@olebole The PR could evolve in that direction. I'm assuming that we still want to derive ERFA (which may contain this automatic functionality) from SOFA. At the moment, I'm just trying to see how big of a footprint ERFA would need to add to SOFA to make it work. And maybe, part of this footprint could be accepted by SOFA directly, making the derivation easier.

**Comment by mhvk** [comment]
@jwoillez - Just so I understand correctly, ideally SOFA would implement the definition of `NDAT` and `changes` outside of the function, as well as the internal setting of it, so that for ERFA it is simple to add the `UpdateLeapSeconds` function to update these numbers.

I wonder about @olebole's point though: in the end one would like some degree of automatic checking and updating whenever `eraDat` gets called. Is your idea that this could then be hidden in `UpdateLeapSeconds`? But if so, we'll eventually need to make an additional change that that routine gets called whenever `eraDat` is called (at least with a date that is beyond the last year in the file and the last time things were checked). I wonder if at that point we are not better off with something closer to your original suggestion: if SOFA makes the variables global, a relatively minimal change we can make in ERFA is to rename `eraDat` to `_eraDat` and add a new `eraDat` that does the check and possible update before passing on to `_eraDat`.

@olebole - just so I understand the Debian situation correctly - it seems that in Debian SOFA has generally been replaced with ERFA? At least, I see that ds9 via tksao depends on libstarlink-ast which in turn depends on liberfa1 (and I don't see the SOFA library). So, if we can have some auto-update mechanism in erfa, this would indeed help multiple packages?

**Comment by olebole** [comment]
@mhvk Yes, this is my point. I replaced SOFA with ERFA when it was available, and in Debian Buster SOFA will be removed from Debian completely. And an external table would help me much -- I would trigger it by the update of the `zoneinfo/leap-seconds.list` table (and write a small converter script to convert that file into whatever is needed for ERFA). ERFA itself is used (transitively) by astropy, Starlink PAL, AST, saotk, and ds9, so indeed this would help to keep multiple packages updated.

**Comment by mhvk** [comment]
OK, so it seems that this PR does move in the right direction, in that it exposes a way to update the information. If we had a follow-up PR with an ERFA-only routine that reads the numbers from a file whenever `eraDat` is first called, then things should work just as you would want to. But, @jwoillez, that does mean we have to be sure that on first call, we already allow an override of the built-in constants.

**Comment by eteq** [comment]
@olebole - just so we're clear: are you saying you do not want it to be *possible* for downstream packages to do their own updates, or that `erfa` should take responsibility for it to some extent but leave open the option for higher-level code to also do it?

I ask partly because some of the astropy machinery will want to allow a python level "update the leap seconds" option, which of course shouldn't go mucking about with a system-level leap seconds file.

**Comment by olebole** [comment]
I see a value in a downstream package maintainable leap second table (f.e. by astropy). However, I think that relying only on that will increase the problems: then, every package using erfa would need to take care about the update of the leap seconds himself. I think a better solution would be that erfa would maintain this by a local table as primary source. Astropy can overwrite this, but users shall not need to rely on astropy to have up-to-date information.
I see this however with my Debian hat on, since I would just keep this up-to-date by a small helper program (which would work on other Linuxes as well), and all users of ERFA would profit automatically.
But my preferred way would probably not work that well for "more chaotic" installations (f.e. erfa as locally integrated part of a pip-Astropy), and I also don't know whether and where leap second information is available and kept up-to-date on MacOS and Windows. That's why I can't just make a general pull request.

**Comment by eteq** [comment]
As @mhvk  mentioned in https://github.com/astropy/astropy/issues/5783#issuecomment-538522998 , as a way to move forward with this (driven by the impending Astropy v4.0 feature freeze), here's what we propose:

* We will attempt to address astropy/astropy#5783 by taking this PR and integrating it into the astropy-bundled erfa (and associated Python wrappers).
* We will then try to use the astropy machinery to update the leap seconds using this functionality as part of the IERS download.
* Assuming this actually works as we hope, we will push back to this PR any fixes that are necessary to make things work as expected in Astropy
* Once that is done, we will aim to merge this PR, but then aim for a follow-on PR that implements the approach @olebole suggests of using local information when it's available.

So basically the proposal here is to use Astropy as a test balloon (since we can paper over some of it on the Python side when necessary), but try to push things back to `liberfa` as soon as we've got it working.  Sound reasonable, @olebole @jwoillez, @mhvk?

**Comment by jwoillez** [comment]
I have to admit that I did not pay attention to this thread for quite a while. The proposed plan looks great.

For the idea of making ERFA carry out the update by itself, based on a leap second file provided by a distribution, one could add a hook where the NDAT and changes structures are initialized. I will annotate the PR to show where.

File: src/dat.c (hunk 5 above)
Comment (jwoillez) [reply src/dat.c]: This is where one may want to had a hook to read and parse a distribution provided leap second file, like `zoneinfo/leap-seconds.list` on Debian. If `NDAT==0`, we are on the first call, and before setting `NDAT` to `_NDAT` and `changes` to `_changes`, one could try to get the update from such files.

**Comment by mhvk** [comment]
Thanks for the comments, @jwolliez - we'll need to think through what to do quite carefully, since ideally we have something with which people can just path SOFA directly (which is now consistent with their license for `dat.c`). For now, more work at https://github.com/astropy/astropy/pull/9329

**Comment by bsipocz** [comment]
@eteq - this needs to be released asap to be bundled properly as a release to astropy v4.0.

**Comment by olebole** [comment]
This patch is not compatible to astropy 4.0rc1; I get an 
```
ImportError: /build/astropy-4.0~rc1/.pybuild/cpython3_3.8/build/astropy/_erfa/ufunc.cpython-38-x86_64-linux-gnu.so: undefined symbol: eraSetLeapSeconds
```
when trying to use it (among some warnings when building).

**Comment by mhvk** [comment]
I just realized I made a different change in astropy itself (based on this, but more separated out to a new file). I'll make a PR shortly.

**Comment by mhvk** [comment]
closing in favour of #58

**Comment by eteq** [comment]
Since it turned out more straightforward to go from #58 (which includes this as its base commit), I'll go ahead and close this and replace the last bullet point from https://github.com/liberfa/erfa/pull/49#issuecomment-538585368 with some new issues.  Thanks @jwoillez !
